[Core] Build time error due to `-Wignored-attributes`

### What happened + What you expected to happen

I'm building Ray from source; the build fails with the following error:

```
src/ray/util/process.cc: In static member function 'static std::string ray::Process::Exec(std::string)':
src/ray/util/process.cc:410:42: error: ignoring attributes on template argument 'int (*)(FILE*)' [-Werror=ignored-attributes]
  410 |   std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(command.c_str(), "r"), pclose);
      |                                          ^
cc1plus: all warnings being treated as errors
```

I'm guessing this is because I'm building with a new version of GCC - I know in the past we've been restricted to much older versions, though I'm surprised we haven't seen this come up as I know it was introduced first in GCC 6 (maybe we're still restricted to gcc<6? As far as I can tell [there's no restriction in the docs](https://docs.ray.io/en/latest/ray-contribute/development.html)).

I'm happy to make a PR for this, in fact I have a branch with the fix already - but I wanted to create an issue to track this. Let me know if a PR would be of interest.

### Versions / Dependencies

`python -VV`: Python 3.11.9 (main, Jul  5 2024, 12:36:19) [GCC 14.1.1 20240522]
`uname -rms`: Linux 6.9.10-arch1-1 x86_64
`gcc -v`:

```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/14.1.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --enable-languages=ada,c,c++,d,fortran,go,lto,m2,objc,obj-c++,rust --enable-bootstrap --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://gitlab.archlinux.org/archlinux/packaging/packages/gcc/-/issues --with-build-config=bootstrap-lto --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-libstdcxx-backtrace --enable-link-serialization=1 --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-werror
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 14.1.1 20240522 (GCC) 
```

Ray version: ff789b95b8bd22af257fc168a9cc8ab689af6ede

### Reproduction script

I actually cannot build `ray` in my current `pyenv` environment using the steps detailed in the documentation (using `pip install -ve .` inside the `ray/python` directory). However I was able to get the compiled extensions building with `python setup.py build_ext` (deprecated, I know, but somehow this was the only way I found to get this building, and anyway the issue I ran into is purely in the C compilation step so I think it's unrelated to my `python setup.py build_ext` invocation which IMO is a separate issue).

### Issue Severity

High: It blocks me from completing my task.

--- a/src/ray/util/process.cc
+++ b/src/ray/util/process.cc
@@ -407,8 +407,7 @@ std::string Process::Exec(const std::string command) {
 #ifdef _WIN32
   std::unique_ptr<FILE, decltype(&_pclose)> pipe(_popen(command.c_str(), "r"), _pclose);
 #else
-  std::unique_ptr<FILE, void (*)(FILE *)> pipe(popen(command.c_str(), "r"),
-                                               [](FILE *f) -> void { pclose(f); });
+  std::unique_ptr<FILE, int (*)(FILE *)> pipe(popen(command.c_str(), "r"), pclose);
 #endif
   RAY_CHECK(pipe) << "popen() failed for command: " + command;
   while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {

--- a/src/ray/util/process.cc
+++ b/src/ray/util/process.cc
@@ -407,7 +407,8 @@ std::string Process::Exec(const std::string command) {
 #ifdef _WIN32
   std::unique_ptr<FILE, decltype(&_pclose)> pipe(_popen(command.c_str(), "r"), _pclose);
 #else
-  std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(command.c_str(), "r"), pclose);
+  std::unique_ptr<FILE, void (*)(FILE *)> pipe(popen(command.c_str(), "r"),
+                                               [](FILE *f) -> void { pclose(f); });
 #endif
   RAY_CHECK(pipe) << "popen() failed for command: " + command;
   while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {

